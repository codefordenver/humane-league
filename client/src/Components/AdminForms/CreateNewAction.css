.create-new-action-container {
  display: flex;
  flex-direction: column;
  background: rgba(0, 0, 0, 0.5);
  width: 100%;
  min-height: calc(100vh - 60px);
  padding: 40px 0px; }
  .create-new-action-container h1 {
    color: black;
    font-size: 24pt;
    text-align: center;
    margin: 20px; }
    .create-new-action-container h1.success {
      color: black; }
    .create-new-action-container h1 span {
      font-size: 28pt;
      font-weight: 700;
      margin: 0 5px; }
  .create-new-action-container .select-action-container {
    background-color: #ffffffe3;
    border-radius: 5px;
    box-shadow: 10px 10px 15px rgba(0, 0, 0, 0.6);
    margin: 0px auto;
    position: relative;
    width: 650px; }
    .create-new-action-container .select-action-container .close-Modal {
      position: absolute;
      right: -20px;
      top: -20px;
      width: 40px;
      height: 40px;
      cursor: pointer;
      background-image: url("../../closeModal.svg");
      transition: 0.5s; }
      .create-new-action-container .select-action-container .close-Modal:hover {
        background-image: url("../../closeModal-hover.svg");
        transform: rotate(90deg); }
    .create-new-action-container .select-action-container label {
      display: flex;
      align-items: center;
      font-size: 20px;
      width: 328px;
      margin: 0 auto; }
    .create-new-action-container .select-action-container select,
    .create-new-action-container .select-action-container option {
      cursor: pointer;
      font-size: 20px;
      height: 40px;
      padding: 5px;
      margin: 20px; }
