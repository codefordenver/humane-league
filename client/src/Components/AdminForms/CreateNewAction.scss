.create-new-action-container {
  display: flex;
  flex-direction: column;
  h1 {
    color: #fff;
    font-size: 24pt;
    text-align: center;
    margin: 20px;
    span {
      font-size: 28pt;
      margin: 0 5px;
    }
  }
  .select-action-container {
    background-color: #ffffffe3;
    border-radius: 2px;
    box-shadow: 2px 2px 15px rgba(0, 0, 0, 0.6);
    margin: 20px;
    label {
      display: block;
      font-size: 20px;
      width: 328px;
      margin: 0 auto;
    }
    select,
    option {
      cursor: pointer;
      font-size: 20px;
      height: 40px;
      padding: 5px;
      margin: 20px;
    }
  }
}