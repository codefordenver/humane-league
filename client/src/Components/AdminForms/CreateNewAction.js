import React, { Component } from 'react';
import { connect } from 'react-redux';
import './AdminForms.css';

class CreateNewAction extends Component {
  constructor() {
    super();
    this.state = {
      form: 'facebook',
      actionEnabled: true,
      error: false,
      success: false
    }
  }

  createAction = (type) => {
    const baseAction = {
      title: this.actionTitle.value,
      description: this.actionDescription.value
    };

    if (type === 'social') {
      const target = this.targetUrl.value;

      return Object.assign({ enabled: this.state.actionEnabled }, baseAction, { target });
    } 
    else if (type === 'email') {
      const email = {
        to: this.emailTo.value,
        cc: this.emailCC.value,
        bcc: this.emailBCC.value,
        subject: this.emailSubject.value
      };

      return Object.assign({ enabled: this.state.actionEnabled }, baseAction, email);
    } 
    else if (type === 'phone') {
      const phone = {
        name: this.phoneName.value,
        position: this.phonePosition.value,
        phone_number: this.phoneNumber.value
      };

      return Object.assign({ enabled: this.state.actionEnabled }, baseAction, phone);
    }
  }

  actionPost = async (action, content, type) => {
    const token = this.props.user.id_token;
    const actionPost = await fetch(`/api/v1/${type}_actions?token=${token}`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(action)
    });
    const actionID = await actionPost.json();

    if (actionID.id) {
      const contentPost = await fetch(`/api/v1/${type}_contents?token=${token}`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify({ action_id: actionID.id, content })
      });
      const contentID = await contentPost.json();
      
      if(contentID.error) {
        this.setState({ error: `Could not create action content: ${contentID.error}`});
        setTimeout(() => {
        this.setState({ error: false });
      }, 5000);
      } else if(contentID.id) {
        return contentID
      }

    } else if (actionID.error) {
      this.setState({ error: `Could not create action: ${actionID.error}`});
      setTimeout(() => {
        this.setState({ error: false });
      }, 5000);
    }
  }

  handleChange = () => {
    const form = this.actionTypes.value;
    
    this.setState({ form });
  }


  handleSubmit = async (event) => {
    event.preventDefault();
    const type = (this.state.form === 'facebook' || this.state.form === 'twitter')
      ? 'social'
      :  this.state.form;
    const action = this.createAction(type);
    const actionContent = this.actionContent.value;
    const postResult = await this.actionPost(action, actionContent, this.state.form);

    if (postResult.id) {
      this.resetForm(type);
      this.setState({ success: 'ACTION CREATED!' });
      setTimeout(() => {
        this.setState({ success: false });
      }, 5000);
    }
  }

  resetForm = (type) => {
    this.actionTitle.value = '';
    this.actionDescription.value = '';
    this.actionContent.value = '';

    if (type === 'social') {
      this.targetUrl.value = '';
    } else if (type === 'email') {
      this.emailTo.value = '';
      this.emailCC.value = '';
      this.emailBCC.value = '';
      this.emailSubject.value = '';
    } else if (type === 'phone') {
      this.phoneName.value = '';
      this.phonePosition.value = '';
      this.phoneNumber.value = '';
    }

    this.setState({ actionEnabled: true });
  }

  render() {
    const socialMediaTarget = {
      targetUrl: <input type='text' ref={(elem) => {this.targetUrl = elem}} placeholder='Target Url' />
    }  
    const emailTarget = {
      to: <input type='text' ref={(elem) => {this.emailTo = elem}} placeholder='Target Email Address' />,
      cc: <input type='text' ref={(elem) => {this.emailCC = elem}} placeholder='CCs' />,
      bcc: <input type='text' ref={(elem) => {this.emailBCC = elem}} placeholder='BCCs' />,
      subject: <input type='text' ref={(elem) => {this.emailSubject = elem}} placeholder='Email Subject' />,
    }
    const phoneTarget = {
      name: <input type='text' ref={(elem) => {this.phoneName = elem}} placeholder='Target Name' />,
      position: <input type='text' ref={(elem) => {this.phonePosition = elem}} placeholder='Target Position' />,
      phoneNumber: <input type='text' ref={(elem) => {this.phoneNumber = elem}} placeholder='Target Phone Number' />
    }
    const form = {
      facebook: socialMediaTarget,
      twitter: socialMediaTarget,
      email: emailTarget,
      phone: phoneTarget
    }

    return (
      <div className='create-new-action-container'>
        <h1>CREATE A NEW <span>{this.state.form.toUpperCase()}</span> ACTION</h1>
        <section className='select-action-container'>
          <label htmlFor='action-types'>Select Action Type:
            <select onChange={() => this.handleChange()} ref={(elem) => {this.actionTypes = elem}} name='action-types' id='action-types'>
              <option value='facebook'>Facebook</option>
              <option value='twitter'>Twitter</option>
              <option value='email'>Email</option>
              <option value='phone'>Phone</option>
            </select>
            {
              this.state.error && 
              <h1>{this.state.error}</h1>
            }
            {
              this.state.success &&
              <h1>{this.state.success}</h1>
            }
          </label>
          <section className='create-action-form'>
            <form action='create-new-action-form'>
              <input type='text' ref={(elem) => {this.actionTitle = elem}} placeholder='Action Title' />
              <input type='text' ref={(elem) => {this.actionDescription = elem}} placeholder='Action Description' />
              {form[this.state.form].targetUrl}

              {form[this.state.form].to}
              {form[this.state.form].cc}
              {form[this.state.form].bcc}
              {form[this.state.form].subject}

              {form[this.state.form].name}
              {form[this.state.form].position}
              {form[this.state.form].phoneNumber}
              <textarea ref={(elem) => {this.actionContent = elem}} placeholder='Action Content'></textarea>

              <span id='toggle'>
                <input
                  type='checkbox' 
                  checked={this.state.actionEnabled}
                  onChange={() => this.setState({ actionEnabled: !this.state.actionEnabled })} />
                <label 
                  data-on='enabled' 
                  data-off='disabled'>
                </label>
              </span>    
              <button onClick={this.handleSubmit}>CREATE THIS ACTION</button>
            </form>
          </section>
        </section>
      </div>
    )
  }
}

const mapStateToProps = store => ({
  user: store.User
});

export default connect(mapStateToProps, null)(CreateNewAction);