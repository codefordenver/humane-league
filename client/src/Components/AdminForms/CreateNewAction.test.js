import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import { CreateNewAction } from './CreateNewAction';

describe('CreateNewAction component tests', () => {
  let renderedCreateNewAction;
  let defaultState = {
    "error": false, "form": "facebook", "success": false
  }
  
  beforeEach(() => {
    renderedCreateNewAction = shallow( <CreateNewAction /> );
  })
  it('renders without crashing', () => {
    expect(renderedCreateNewAction).toBeDefined();
    expect(renderedCreateNewAction.state()).toEqual(defaultState);
  });
  it('matches SnapShot', () => {
    expect(renderedCreateNewAction).toMatchSnapshot();
  })
})