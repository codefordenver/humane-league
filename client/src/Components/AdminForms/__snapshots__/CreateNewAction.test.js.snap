// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateNewAction component tests matches SnapShot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CreateNewAction />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1>
          CREATE A NEW 
          <span>
                    FACEBOOK
          </span>
           ACTION
</h1>,
        <section
          className="select-action-container"
>
          <label
                    htmlFor="action-types"
          >
                    Select Action Type:
                    <select
                              id="action-types"
                              name="action-types"
                              onChange={[Function]}
                    >
                              <option
                                        value="facebook"
                              >
                                        Facebook
                              </option>
                              <option
                                        value="twitter"
                              >
                                        Twitter
                              </option>
                              <option
                                        value="email"
                              >
                                        Email
                              </option>
                              <option
                                        value="phone"
                              >
                                        Phone
                              </option>
                    </select>
          </label>
          <ActionForm
                    form="facebook"
                    handleSubmit={[Function]}
          />
</section>,
      ],
      "className": "create-new-action-container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "CREATE A NEW ",
            <span>
              FACEBOOK
</span>,
            " ACTION",
          ],
        },
        "ref": null,
        "rendered": Array [
          "CREATE A NEW ",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "FACEBOOK",
            },
            "ref": null,
            "rendered": "FACEBOOK",
            "type": "span",
          },
          " ACTION",
        ],
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <label
              htmlFor="action-types"
>
              Select Action Type:
              <select
                            id="action-types"
                            name="action-types"
                            onChange={[Function]}
              >
                            <option
                                          value="facebook"
                            >
                                          Facebook
                            </option>
                            <option
                                          value="twitter"
                            >
                                          Twitter
                            </option>
                            <option
                                          value="email"
                            >
                                          Email
                            </option>
                            <option
                                          value="phone"
                            >
                                          Phone
                            </option>
              </select>
</label>,
            <ActionForm
              form="facebook"
              handleSubmit={[Function]}
/>,
          ],
          "className": "select-action-container",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "Select Action Type:",
                <select
                  id="action-types"
                  name="action-types"
                  onChange={[Function]}
>
                  <option
                                    value="facebook"
                  >
                                    Facebook
                  </option>
                  <option
                                    value="twitter"
                  >
                                    Twitter
                  </option>
                  <option
                                    value="email"
                  >
                                    Email
                  </option>
                  <option
                                    value="phone"
                  >
                                    Phone
                  </option>
</select>,
                false,
                false,
              ],
              "htmlFor": "action-types",
            },
            "ref": null,
            "rendered": Array [
              "Select Action Type:",
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <option
                      value="facebook"
>
                      Facebook
</option>,
                    <option
                      value="twitter"
>
                      Twitter
</option>,
                    <option
                      value="email"
>
                      Email
</option>,
                    <option
                      value="phone"
>
                      Phone
</option>,
                  ],
                  "id": "action-types",
                  "name": "action-types",
                  "onChange": [Function],
                },
                "ref": [Function],
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "Facebook",
                      "value": "facebook",
                    },
                    "ref": null,
                    "rendered": "Facebook",
                    "type": "option",
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "Twitter",
                      "value": "twitter",
                    },
                    "ref": null,
                    "rendered": "Twitter",
                    "type": "option",
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "Email",
                      "value": "email",
                    },
                    "ref": null,
                    "rendered": "Email",
                    "type": "option",
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "Phone",
                      "value": "phone",
                    },
                    "ref": null,
                    "rendered": "Phone",
                    "type": "option",
                  },
                ],
                "type": "select",
              },
              false,
              false,
            ],
            "type": "label",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "form": "facebook",
              "handleSubmit": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "section",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1>
            CREATE A NEW 
            <span>
                        FACEBOOK
            </span>
             ACTION
</h1>,
          <section
            className="select-action-container"
>
            <label
                        htmlFor="action-types"
            >
                        Select Action Type:
                        <select
                                    id="action-types"
                                    name="action-types"
                                    onChange={[Function]}
                        >
                                    <option
                                                value="facebook"
                                    >
                                                Facebook
                                    </option>
                                    <option
                                                value="twitter"
                                    >
                                                Twitter
                                    </option>
                                    <option
                                                value="email"
                                    >
                                                Email
                                    </option>
                                    <option
                                                value="phone"
                                    >
                                                Phone
                                    </option>
                        </select>
            </label>
            <ActionForm
                        form="facebook"
                        handleSubmit={[Function]}
            />
</section>,
        ],
        "className": "create-new-action-container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "CREATE A NEW ",
              <span>
                FACEBOOK
</span>,
              " ACTION",
            ],
          },
          "ref": null,
          "rendered": Array [
            "CREATE A NEW ",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "FACEBOOK",
              },
              "ref": null,
              "rendered": "FACEBOOK",
              "type": "span",
            },
            " ACTION",
          ],
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <label
                htmlFor="action-types"
>
                Select Action Type:
                <select
                                id="action-types"
                                name="action-types"
                                onChange={[Function]}
                >
                                <option
                                                value="facebook"
                                >
                                                Facebook
                                </option>
                                <option
                                                value="twitter"
                                >
                                                Twitter
                                </option>
                                <option
                                                value="email"
                                >
                                                Email
                                </option>
                                <option
                                                value="phone"
                                >
                                                Phone
                                </option>
                </select>
</label>,
              <ActionForm
                form="facebook"
                handleSubmit={[Function]}
/>,
            ],
            "className": "select-action-container",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "Select Action Type:",
                  <select
                    id="action-types"
                    name="action-types"
                    onChange={[Function]}
>
                    <option
                                        value="facebook"
                    >
                                        Facebook
                    </option>
                    <option
                                        value="twitter"
                    >
                                        Twitter
                    </option>
                    <option
                                        value="email"
                    >
                                        Email
                    </option>
                    <option
                                        value="phone"
                    >
                                        Phone
                    </option>
</select>,
                  false,
                  false,
                ],
                "htmlFor": "action-types",
              },
              "ref": null,
              "rendered": Array [
                "Select Action Type:",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      <option
                        value="facebook"
>
                        Facebook
</option>,
                      <option
                        value="twitter"
>
                        Twitter
</option>,
                      <option
                        value="email"
>
                        Email
</option>,
                      <option
                        value="phone"
>
                        Phone
</option>,
                    ],
                    "id": "action-types",
                    "name": "action-types",
                    "onChange": [Function],
                  },
                  "ref": [Function],
                  "rendered": Array [
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": "Facebook",
                        "value": "facebook",
                      },
                      "ref": null,
                      "rendered": "Facebook",
                      "type": "option",
                    },
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": "Twitter",
                        "value": "twitter",
                      },
                      "ref": null,
                      "rendered": "Twitter",
                      "type": "option",
                    },
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": "Email",
                        "value": "email",
                      },
                      "ref": null,
                      "rendered": "Email",
                      "type": "option",
                    },
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": "Phone",
                        "value": "phone",
                      },
                      "ref": null,
                      "rendered": "Phone",
                      "type": "option",
                    },
                  ],
                  "type": "select",
                },
                false,
                false,
              ],
              "type": "label",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "form": "facebook",
                "handleSubmit": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "section",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
